### DISCUSSION TITLE
Numbers in amounts

### DISCUSSION THREAD

**Discussion by Accacio** [cooklang/spec#33] [General]
I have a question: What do we gain in dividing numbers into multiple types? couldn't we just get an expression and calculate the resulting values? This way people could even make operations in the quantities and 1/2 would be parsed as .5 in float/double.

https://github.com/cooklang/spec/blob/5700a66171981e324188dc57efebbb0d1dee5739/EBNF.md?plain=1#L36-L42

**Comment by dubadub** [comment]
We can and probably it won't be a big deal. I kept them this way because of rounding errors. `1/2` is easy but with `1/3` or `0.1` the error will creep in. I delay the conversion until the end, so then I can decide during formatting how to handle the best way.

```
Welcome to Swift version 5.5.1-dev.
Type :help for assistance.
  1> 0.1
$R0: Double = 0.10000000000000001
```

**Reply by dubadub** [reply]
@Accacio eventually I was convinced that having common double/float is a better option, I wasn’t right in the beginning. In canonical tests I used that approach https://github.com/cooklang/spec/blob/main/tests/canonical.yaml#L164. I don’t think that’s relevant to EBNF though and no changes should be made.

**Reply by heytherewill** [reply]
If the quantities should be floating point numbers, why do some canonical tests have [strings](https://github.com/cooklang/spec/blob/main/tests/canonical.yaml#L256) and the [raw fractions](https://github.com/cooklang/spec/blob/main/tests/canonical.yaml#L228) as valid parsing results for quantities?

**Reply by heytherewill** [reply]
I ask because I'm currently working on making my parser pass the canonical tests and this is the final bit missing. Since I'm using a statically typed language, I'd have to model this as a string (which is far from ideal, since a quantity should be a quantity). I just wanna understand what the resulting values should be, an from the looks of it, they should be floating point numbers but there are some outliers in the tests which should be changed

**Reply by Accacio** [reply]
After parsed the values should be floats, 3/2 should become  1.5

**Reply by heytherewill** [reply]
We agree on that, but the linked examples in the canonical tests show otherwise (an instance of fraction being parsed as a string and another where the quantity is not a number at all).

What I'm saying is that the canonical tests should reflect that correctly by changing that `"some"` and `01/2`

**Reply by Accacio** [reply]
ah ok, sorry. My fault. I didn't see the links when I opened your comment on the mobile, I've seen the examples and you are right.

**Reply by dubadub** [reply]
> If the quantities should be floating point numbers, why do some canonical tests have [strings](https://github.com/cooklang/spec/blob/main/tests/canonical.yaml#L256) and the [raw fractions](https://github.com/cooklang/spec/blob/main/tests/canonical.yaml#L228) as valid parsing results for quantities?

Quantities can be either number or string as per [EBNF](https://github.com/cooklang/spec/blob/main/EBNF.md?plain=1#L27). You can achieve this with a statically typed language as well, you might take a look at [swift parser](https://github.com/cooklang/cooklang-swift/blob/main/Sources/CookInSwift/Parser/AST.swift#L13-L17).
